janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexHasUniquePropertyOptimizerStrategy.java

```diff
@@ -0,0 +1,146 @@
+
+        List<PredicateCondition<String, JanusGraphElement>> constraints = step.getHasContainers().stream()
+            .filter(hc -> hc.getBiPredicate() == Compare.eq)
+            .map(hc -> new PredicateCondition<>(hc.getKey(), JanusGraphPredicateUtils.convert(hc.getBiPredicate()), hc.getValue()))
+            .collect(Collectors.toList());
+        final MultiCondition<JanusGraphElement> conditions = QueryUtil.constraints2QNF(tx, constraints);
+
+        // collect all matching indexes
+        final Set<IndexType> availableIndexes = new HashSet<>();
+        ConditionUtil.traversal(conditions, condition -> {
+            if (condition instanceof PredicateCondition) {
+                final RelationType type = ((PredicateCondition<RelationType,JanusGraphElement>) condition).getKey();
+                Preconditions.checkArgument(type != null && type.isPropertyKey());
+                Iterables.addAll(availableIndexes, ((InternalRelationType) type).getKeyIndexes());
+            }
+            return true;
+        });
+
+        // select only valid unique indexes
+        return availableIndexes.stream()
+            .filter(IndexType::isCompositeIndex)
+            .map(index -> (CompositeIndexType) index)
+            .filter(cIndex -> cIndex.getCardinality() == Cardinality.SINGLE)
+            .filter(cIndex -> isIndexSatisfiedByGivenKeys(cIndex, givenKeys))
+            .count() > 0;
+    }
+
+    private boolean isIndexSatisfiedByGivenKeys(IndexType index, Collection<String> givenKeys) {
+        for (int i = 0; i < index.getFieldKeys().length; ++i) {
+            if (!givenKeys.contains(index.getFieldKeys()[i].getFieldKey().name())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private Traversal.Admin<?,Long> generateFilter(Traversal.Admin<?,?> traversal, HasStep<?> originalStep) {
+        Traversal.Admin filterTraversal = new DefaultGraphTraversal<>(traversal.getGraph().get())
+            .addStep(
+                new GraphStep<>(traversal, Vertex.class, true)
+            );
+
+        filterTraversal.addStep(originalStep);
+        filterTraversal.addStep(new IdStep<>(traversal));
+        return filterTraversal;
+    }
+
+    private long evaluateFilter(Traversal.Admin<?,Long> filterTraversal) {
+        // negative IDs are not valid in JanusGraph. Therefore, if the filterTraversal has no result, the newly
+        // inserted HasStep should not match anything
+        long determinedId = -1;
+        if (filterTraversal.hasNext()) {
+            determinedId = (long) filterTraversal.next();
+        }
+        return determinedId;
+    }
+
+    @Override
+    protected FilterStep<Edge> makeFilterByAdjacentIdStep(Traversal.Admin<?, ?> traversal, HasStep<?> originalStep) {
+        Traversal.Admin<?,Long> filterTraversal = generateFilter(traversal, originalStep);
+        HasContainer hc = new HasContainer(ADJACENT_ID.name(), P.eq(evaluateFilter(filterTraversal)));
```

Comment:
Yes, that's what it does. Same question as above: What exactly to you mean by profiling?

Reply:
If you run query.profile(), the latency metrics would be inaccurate because it does not take this query latency into account. By profiling I mean you need to have a timer here, which is a part of the query profile.

Reply:
Alright, I'll try to find out how to add a timer to the query profile. I thought that would happen automatically for optimization strategies because `optimization` appears quite often in the query profiles that I see.

Reply:
Adding custom profiling outside the actual query execution is not intended. The metrics, in which the profile is stored, are created within the first `next()` or `hasNext()` call of a `ProfileStep`. Therefore, there is no profiler available at optimization time.
The only "solution" that comes to my mind is a pretty ugly one and in fact a complete re-write of the current state:
Instead of doing the ID resolution at optimization time, we would have to do it at runtime. To do so, we would need to implement a custom step which holds both the unoptimized query and a skeleton of the optimized query as it's children. Then at runtime, this step could evaluate the index uniqueness and dynamically decide which path to use. It would then also have to inject the resolved ID into the optimized query somehow. In my eyes, this is a mess...

If you have any other idea, please let me know.

Reply:
To do profiling in the optimization step, you can actually create a new `TP3ProfileWrapper` and later in runtime, add the metrics of this "optimization profiler" to the query profiler. I did something similar in another issue I have been working on which I would probably submit a PR in 1-2 weeks.

Actually I am thinking that this ad-hoc traversal shall indeed be done only in runtime. Let's say we have
```groovy
g.V().has("key1", "value1").out(label).has("key2", "value2")
```
If `g.V().has("key1", "value1")` does not exist, the query will finish. It could also be the case that this `g.V().has("key1", "value1")` vertex does not have any `out(label)` edge at all. Under such "early-stopping" circumstances, the ad-hoc traversal done for `has("key2", "value2")` in the optimization stage becomes meaningless and brings only overhead. If this ad-hoc traversal is lazily computed, we don't have such a problem.

That being said, I do also see the potential performance improvement if we have `has("key2", "value2")` in the optimization stage, since it can eliminate the need of the whole traversal when `has("key2", "value2")` returns null. If we don't put this logic in this strategy, then it does not make much sense to have the ad-hoc traversal in optimization stage.

Reply:
As mentioned [earlier](https://github.com/JanusGraph/janusgraph/pull/2189#discussion_r479911325), I would prefer not to have this logic in the same strategy but instead have a "cleanup" strategy which runs at the end of the optimization phase and prunes the query by eliminating all sub queries that contain unsatisfiable condition.
Other optimization strategies (like this one) could then mark conditions as unsatisfiable by adding/replacing with a `hasId(-1)` step for example. This way, we can ensure we don't have to implement a similar logic in every strategy that makes use of it.

Also thanks for pointing me towards `TP3ProfileWrapper`. If you don't mind, I would like to wait until I see how you are handling this in your PR.

Reply:
> I would prefer not to have this logic in the same strategy but instead have a "cleanup" strategy...

That's fair.

> I would like to wait until I see how you are handling this in your PR

Sure 😉

Reply:
It's been a while and after this PR was requested to be merged soon in yesterday's JanusGraph meetup, I came back to look at what's still missing here in order to merge the PR. Throughout the day I tried different ways to integrate the collection of metrics into this optimization strategy. But in contrast to what you did in your PR, the step within the traversal that is created by my optimization strategy, is later merged with previous `JanusGraphVertexStep`s and will therefore not receive it's own query metrics by the query executor.
The problem is that the `has` step created by my strategy is only intermediate and will never remain once the query is completely optimized by the `JanusGraphLocalQueryOptimizerStrategy`. Therefore at the time the initial `Metrics` Object is created, there is no step remaining that can possibly be responsible for accounting the time spent in the optimization phase.

One possible solution would be to create a custom `IdentityStep` within the query that does nothing but hold the metrics from the optimization phase. But this step could possibly interfere with other optimization strategies and get removed by them or make it harder for them to be applied.

Concluding, I don't see a really good solution for this problem. IMHO the cleanest way would be to somehow measure the time for each optimization independently from the query execution, bundle them and somehow integrate them in the profile result afterwards. But this would require a lot of work to be done and design decisions to be made.

Reply:
Quick addition: As some optimizations are already applied by TinkerPop and therefore outside of JanusGraph, it would obviously preferred to have this kind of functionality as part of TinkerPop, not JanusGraph.

Reply:
I yet to find some time to review this, but 
> this PR was requested to be merged soon in yesterday's JanusGraph meetup

this grasps my attention 😄.  I didn't attend the meetup but I am curious about the request: is there a planned date for the 0.6.0 release? If not, why would users (I assume this was a request by some user) think merging this PR soon will help them? Is it because they are using the master version of JanusGraph?

Reply:
I don't think so. But from a user perspective if they neither know when 0.6.0 will be released nor when this feature will be merged, chances of getting it with v0.6.0 are better the earlier this change is in.
If you would like to view the recording of the meetup (where I gave a talk, too :tada:), I think @twilmes said he plans to upload it somewhere it's accessible for those who were not able to make it to the meetup.